llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -4074,14 +4080,62 @@ template <> Value *SPIRVToLLVM::transValueWithOpcode<OpExtInst>(SPIRVValue *cons
 
   case SPIRVEIS_Debug:
     return m_dbgTran.transDebugIntrinsic(spvExtInst, block);
-
+  case SPIRVEIS_NonSemanticDebugPrintf:
+    switch (spvExtInst->getExtOp()) {
+    case NonSemanticDebugPrintfDebugPrintf:
+      return transDebugPrintf(spvExtInst, spvArgValues, func, block);
+    default:
+      return nullptr;
+    }
   default:
     llvm_unreachable("Should never be called!");
     return nullptr;
   }
 }
 
 // =====================================================================================================================
+// Handle extended instruction Nonsemantics.debugPrintf
+//
+// @param bi : A SPIR-V debugPrintf instruction
+// @param spvValues: Instruction spirv parameters
+// @param func: Which function to generate code
+// @param bb: Which basicblock to generate code
+Value *SPIRVToLLVM::transDebugPrintf(SPIRVInstruction *bi, const ArrayRef<SPIRVValue *> spvValues, Function *func,
+                                     BasicBlock *bb) {
+  if (!getBuilder()->findResourceNode(ResourceNodeType::DescriptorBuffer, Vkgc::InternalDescriptorSetId,
+                                      Vkgc::PrintfBufferBindingId)) {
+    return getBuilder()->getInt64(0);
+  }
+  if (!m_debugOutputBuffer) {
+
+    auto spvArrType = m_bm->addRuntimeArray(m_bm->addIntegerType(32));
+    auto spvStructType = m_bm->addStructType({spvArrType});
+    Type *bufType = transType(spvStructType);
+
+    m_debugOutputBuffer =
+        new GlobalVariable(*m_m, bufType, false, GlobalValue::ExternalLinkage, nullptr, "debugOutputBuffer", nullptr,
```

Comment:
Can we use `getBuilder()->CreateGetDescPtr()` here to get the printf buffer descriptor? then you don't need to create the globalVariable and the metadata here.

Reply:
This was discussed in the internally before. Using global variables can entail cheap optimization. E.g.  for several raytracing shaders, and each shader has several debugPrintf.  Global variable can be merged into one after linking, and these numerous debug printf can share the same desc which is lowered in the beginning of the entry point.

Reply:
Sounds reasonable. Could you please add the comment into the source file?